### PR TITLE
Feature/962 Display Agree/Disagree user opinion

### PR DESCRIPTION
**Pull request by emullaney**
References #962 Reviewer consensus (agree/disagree) display
 - Review block on submission detail page: agreement/disagreement shown as 'Agree' or 'Disagree' associated with the parent review 
-- Display all agree/disagree opinions
-- Assigned reviewers who have not submitted a review but HAVE submitted an *opinion* won't appear greyed out in the list (we aren't waiting for their review because they have submitted an opinion)

- Expanded submission on Submissions listing: agreement/disagreement shown in Review outcomes section as 'Agree' or 'Disagree' associated with the parent review. (Image shows location only, no need to match design)
image

- Review detail view shows agreement/disagreement from anyone who has agreed or disagreed with the review. A list of:  `<user> <role icon> <agree|disagree>s` will appear in the header to the right of the title.
-- The code displays agree/disagree data below the title, but this can be adjusted on the FE ticket #979 
-- Created a new include template to get this data `review_opinions_list.html` and a new model method on Review Opinion, `get_author_role`

- View all reviews view shows agreement/disagreement from anyone who has agreed or disagreed with the review. This will be an extra row above the score with a list as per Review Detail page.
-- I reused the include template `review_opinions_list.html` in order to display this data on the all reviews table below each score. I'm curious if you're okay with this approach @todd-dembrey 

- Tests

### CODE REVIEW

File: opentech/apply/review/tests/test_views.py [hunk 1]
```diff
@@ -206,3 +207,65 @@ def test_review_detail_recommendation(self):
...
+
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+        cls.submission = ApplicationSubmissionFactory(status='draft_proposal', workflow_stages=2)
+
+    def get_kwargs(self, instance):
+        return {'pk': instance.id, 'submission_pk': instance.submission.id}
+
+    def test_can_see_opinion_buttons_on_others_review(self):
```
Comment (todd-dembrey) [inline opentech/apply/review/tests/test_views.py]: Test that others cant see the buttons and test they cant submit.

File: opentech/apply/review/views.py [hunk 2]
```diff
@@ -30,32 +34,48 @@ def get_context_data(self, **kwargs):
         for review in reviews:
             reviews_dict[review.author.pk] = review
 
+        # Get all the authors of opinions, these authors should not show up in the 'xxx_not_reviewed' lists
+        opinion_authors = set(User.objects.filter(
```
Comment (todd-dembrey) [inline opentech/apply/review/views.py]: `distinct()` rather than set

File: opentech/apply/review/views.py [hunk 3]
```diff
@@ -30,32 +34,48 @@ def get_context_data(self, **kwargs):
...
+            pk__in=ReviewOpinion.objects.filter(review__submission=self.object).values('author__pk')))
```
Comment (todd-dembrey) [inline opentech/apply/review/views.py]: `pk__in=self.object.reviews.opinions.values('author')` maybe?

Would involve an extra method on the review queryset.

File: opentech/apply/review/views.py [hunk 4]
```diff
@@ -30,32 +34,48 @@ def get_context_data(self, **kwargs):
...
                 if review:
                     key = 'staff_reviewed'
-                else:
+                elif review not in opinion_authors:
                     key = 'staff_not_reviewed'
             else:
                 if review:
                     key = 'external_reviewed'
-                else:
+                elif review not in opinion_authors:
```
Comment (todd-dembrey) [inline opentech/apply/review/views.py]: This shouldn't happen. only staff can leave reviews.

File: opentech/apply/review/models.py [hunk 5]
```diff
@@ -168,3 +168,19 @@ def update_submission_reviewers_list(sender, **kwargs):
...
+        settings.AUTH_USER_MODEL,
+        on_delete=models.PROTECT,
+    )
+    opinion = models.IntegerField(choices=OPINION_CHOICES)
+
+    class Meta:
+        unique_together = ('author', 'review')
+
+    def get_author_role(self):
+        role = self.review.submission.assigned.with_roles().filter(reviewer=self.author).first()
```
Comment (todd-dembrey) [inline opentech/apply/review/models.py]: this is `assignment` not `role`.

File: opentech/apply/review/views.py [hunk 6]
```diff
@@ -30,32 +34,48 @@ def get_context_data(self, **kwargs):
...
+                review_info_dict = {
+                    'reviewer': reviewer,
+                    'review': review,
+                    'role': role,
+                }
+                opinions_list = []
+                if review:
+                    # Loop through all opinions and include the current author's role for this submission
+                    for opinion in review.opinions.all():
+                        opinions_list.append({
```
Comment (todd-dembrey) [inline opentech/apply/review/views.py]: i dont think you need to prefetch this in the view. you have access to all of this information off the review object within the templates.

You can even `selected_related("author")` to reduce queries.

File: opentech/apply/review/views.py [hunk 7]
```diff
@@ -181,6 +270,12 @@ def get_context_data(self, **kwargs):
 
         for i, review in enumerate(self.object_list):
             review_data['title']['answers'].append('<a href="{}">{}</a>'.format(review.get_absolute_url(), review.author))
+            if review.opinions:
```
Comment (todd-dembrey) [inline opentech/apply/review/views.py]: do you need the if statement? Perhaps select related the authors on opinions again.

if there are no opinions could the template return nothing/minimal markup? 

You can use an `empty` template tag to handle this nicely.

File: opentech/apply/review/views.py (hunk 3 above)
Comment (emullaney) [reply opentech/apply/review/views.py]: good idea

File: opentech/apply/review/views.py (hunk 4 above)
Comment (emullaney) [reply opentech/apply/review/views.py]: true, I'll switch this to an `else`

File: opentech/apply/review/models.py (hunk 5 above)
Comment (emullaney) [reply opentech/apply/review/models.py]: I think you're just talking about renaming this method? I'll do that.

File: opentech/apply/review/views.py (hunk 6 above)
Comment (emullaney) [reply opentech/apply/review/views.py]: you are correct, excellent point

**Pull request merged**